returnn/datasets/distrib_files.py

```diff
@@ -245,26 +258,45 @@ def init_seq_order(self, epoch: Optional[int] = None, seq_list=None, seq_order=N
         self._lazy_init_file_sizes()
         self._lazy_init_file_cache_proc()
 
-        # Cleanup and fill _files_order_cache.
         full_epoch_0idx = (epoch - 1) // self.partition_epoch
-        for k in list(self._files_order_cache.keys()):
+
+        # Shard files across GPU workers with deterministic seed
+        for k in list(self._files_for_this_worker.keys()):
             if k < full_epoch_0idx:
-                del self._files_order_cache[k]
+                del self._files_for_this_worker[k]
         for ep_ in range(epoch, epoch + self.preload_next_n_sub_epochs + 1):
             full_epoch_0idx_ = (ep_ - 1) // self.partition_epoch
-            if full_epoch_0idx_ in self._files_order_cache:
+            if full_epoch_0idx_ in self._files_for_this_worker:
                 continue
             if self.seq_ordering == "default":
-                files_order_flat = self.files
+                files = self.files
             elif self.seq_ordering == "random":
-                rnd_seed = self._get_random_seed_for_epoch(full_epoch_0idx_ * self.partition_epoch + 1)
-                random_generator = numpy.random.RandomState(rnd_seed)
-                files_order_flat = list(self.files)
-                random_generator.shuffle(files_order_flat)
+                # do not use `self._get_random_seed_for_epoch` to exclude
+                # `random_seed_offset` to ensure an equal distribution in every
+                # worker since the sequences might collide between workers if a
+                # different seed is chosen.
+                random_generator = numpy.random.RandomState(full_epoch_0idx_ * 9941 + 4423)
```

Comment:
Btw, it is still wrong that you do not use `_get_random_seed_for_epoch` here. This is actually the correct function to use. If `random_seed_offset` is set, it should be used. (What you actually want is to fix `_get_default_random_seed_offset` to also allow sharding, i.e. to not use a different random seed offset per worker.)

Reply:
> Btw, it is still wrong that you do not use _get_random_seed_for_epoch here. This is actually the correct function to use. If random_seed_offset is set, it should be used. (What you actually want is to fix _get_default_random_seed_offset to also allow sharding, i.e. to not use a different random seed offset per worker.)

I'm not sure I understand this completely here. What you say seems contradictionary to me.

It seems right to use the random seed offset for all cases except for this one, so I'm not sure why we should adapt this function. Like all the other dataset shuffling logic in the sub datasets should use the offset, and if we're in a worker process in a multi-GPU setting here loading data through `DistributeFilesDataset` we most probably will already have a random seed offset set (which you said _should_ then be used) based on the local rank.

In that sense the current logic seems right to me? Or do you just want to pass a flag to the function like `use_random_seed_offset` that runs the existing logic but specifically excludes the offset?

Reply:
`_get_random_seed_for_epoch` is supposed to return the random seed for an epoch, in all cases, including the case here. If you want a different random seed offset per worker, then `_get_random_seed_for_epoch` should return that. If you want the same random seed offset per worker, then `_get_random_seed_for_epoch` should return that.

`_get_random_seed_for_epoch` should not just always return a different random seed per worker. That would be incorrect behavior of `_get_random_seed_for_epoch`.

`_get_random_seed_for_epoch` can also do other logic. E.g. there is `fixed_random_seed`. In the future, there might be other things. We should always use this function we need the random seed for the epoch.

If `_get_random_seed_for_epoch` does not return the correct random seed for the epoch, then you should consider this a bug of `_get_random_seed_for_epoch`. And you are not using it here because apparently it does not return the correct random seed for the epoch.

So, why do you not use `_get_random_seed_for_epoch`? Why is it not the correct seed? This is because of wrong/buggy/incomplete code in `_get_default_random_seed_offset`. As said before, `_get_default_random_seed_offset` handles both cases correctly in case for TF already, just not for PT.

Reply:
My point is that even in the correctly configured TF case where `random_seed_offset` is used, we _still do not want_ the output of `_get_default_random_seed_offset` because we do not want to use it, but we _do_ want to use it in the sub datasets. It's an entirely separate case, no?

Reply:
Of course we always want `_get_default_random_seed_offset`? Why do you think we would not want it? `_get_default_random_seed_offset` should just return the correct random seed offset depending on what behavior you want.

Reply:
Very much the same argument for `_get_default_random_seed_offset`: `_get_default_random_seed_offset` should not always return a different random seed offset per worker. It should always return the correct random seed offset instead.

Reply:
But we only want this very specific behavior in this one dataset. And `_get_default_random_seed_offset ` is static, so it cannot be overridden. I find it strange to special case that generic implementation there just for this dataset?

Reply:
No it's not only for this dataset. It's always when you want to have the same random seed offset per worker.

But you are right that we should discuss on how we should handle this in general. For TF/Horovod, it checks `is_dataset_distribution_random_seed_offset`. This is somewhat a global setting. As I said before, maybe we should have the same for PT. Or maybe instead of having such a global setting, it should be part of the base `Dataset` class? I'm not sure on this.

In any case, if `random_seed_offset` is explicitly set by the user, or by a parent dataset, or when `RANDOM_SEED_OFFSET_ENV_VAR` is set, it would be used over any such logic. This should basically always be the case when there is a parent dataset. So we are really only discussing the case for the top datasets here.

Reply:
Do you agree to merge this PR and decide on a better solution for the seed in a follow-up?

Reply:
I think this should be done within this PR, as this is actually one of the core aspects of the PR.

Reply:
In any case, the fix would also be very simple. We just should decide on what solution we prefer. Another idea: A dataset can specify whether it supports and uses its own custom sharding in distributed training (like in this case), so some function `Dataset._uses_custom_distributed_sharding`, which would be overwritten by the dataset here, and the `Dataset._get_default_random_seed_offset` would check for it.